### ISSUE TITLE
CSV/XLSX/PDF report generation fails

### ISSUE THREAD

**Issue by DigitalLeaves** [cattr-app/backend-application#4]
Environment:
- OS: Linux Debian 10
- Node 12.16
- MySQL 8.13
- PHP 7.4
- Apache 2.4

When I am in the screen "Dashboard" and I click on the download report icon at the right, choosing any option (CSV, XLSX, PDF) throws an error

![Screenshot 2020-05-24 at 13 09 29](https://user-images.githubusercontent.com/5340535/82751531-8f333180-9dc0-11ea-9e78-3b7641733921.jpg)

The Apache log.
```
XXX.XXX.XXX.XXX - - [24/May/2020:10:09:25 +0000] "POST /api/v1/reports/dashboard HTTP/1.1" 404 978 "https://myhost.com/dashboard/team?type=day&start=2020-05-24&end=2020-05-24" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36"
```

The console:
```
chunk-vendors~d2305125.44ecf14d.js:7 Error: Request failed with status code 404
    at t.exports (chunk-vendors~2a42e354.64b18aae.js:1)
    at t.exports (chunk-vendors~2a42e354.64b18aae.js:3)
    at XMLHttpRequest.h.onreadystatechange (chunk-vendors~2a42e354.64b18aae.js:3)
ie @ chunk-vendors~d2305125.44ecf14d.js:7
re @ chunk-vendors~d2305125.44ecf14d.js:7
ee @ chunk-vendors~d2305125.44ecf14d.js:7
(anonymous) @ chunk-vendors~d2305125.44ecf14d.js:7
Promise.then (async)
ne @ chunk-vendors~d2305125.44ecf14d.js:7
n @ chunk-vendors~d2305125.44ecf14d.js:7
ne @ chunk-vendors~d2305125.44ecf14d.js:7
En.t.$emit @ chunk-vendors~d2305125.44ecf14d.js:7
onExport @ dashboard~748942c6.ac471301.js:1
ne @ chunk-vendors~d2305125.44ecf14d.js:7
n @ chunk-vendors~d2305125.44ecf14d.js:7
ne @ chunk-vendors~d2305125.44ecf14d.js:7
En.t.$emit @ chunk-vendors~d2305125.44ecf14d.js:7
handleMenuItemClick @ chunk-vendors~2a42e354.64b18aae.js:3
ne @ chunk-vendors~d2305125.44ecf14d.js:7
En.t.$emit @ chunk-vendors~d2305125.44ecf14d.js:7
dispatch @ chunk-vendors~2a42e354.64b18aae.js:3
handleClick @ chunk-vendors~2a42e354.64b18aae.js:3
click @ chunk-vendors~2a42e354.64b18aae.js:3
ne @ chunk-vendors~d2305125.44ecf14d.js:7
n @ chunk-vendors~d2305125.44ecf14d.js:7
Qi.o._wrapper @ chunk-vendors~d2305125.44ecf14d.js:7
chunk-vendors~2a42e354.64b18aae.js:3 POST https://myhost.com/api/v1/reports/dashboard 404 (Not Found)
(anonymous) @ chunk-vendors~2a42e354.64b18aae.js:3
t.exports @ chunk-vendors~2a42e354.64b18aae.js:3
t.exports @ chunk-vendors~2a42e354.64b18aae.js:3
Promise.then (async)
u.request @ chunk-vendors~2a42e354.64b18aae.js:1
i.forEach.u.<computed> @ chunk-vendors~2a42e354.64b18aae.js:1
(anonymous) @ chunk-vendors~2a42e354.64b18aae.js:1
value @ dashboard~748942c6.ac471301.js:1
(anonymous) @ dashboard~748942c6.ac471301.js:1
s @ chunk-vendors~b58f7129.83931357.js:1
(anonymous) @ chunk-vendors~b58f7129.83931357.js:1
L.forEach.t.<computed> @ chunk-vendors~b58f7129.83931357.js:1
r @ chunk-vendors~f4520c04.6f36e2e8.js:1
c @ chunk-vendors~f4520c04.6f36e2e8.js:1
(anonymous) @ chunk-vendors~f4520c04.6f36e2e8.js:1
(anonymous) @ chunk-vendors~f4520c04.6f36e2e8.js:1
onExport @ dashboard~748942c6.ac471301.js:1
ne @ chunk-vendors~d2305125.44ecf14d.js:7
n @ chunk-vendors~d2305125.44ecf14d.js:7
ne @ chunk-vendors~d2305125.44ecf14d.js:7
En.t.$emit @ chunk-vendors~d2305125.44ecf14d.js:7
onExport @ dashboard~748942c6.ac471301.js:1
ne @ chunk-vendors~d2305125.44ecf14d.js:7
n @ chunk-vendors~d2305125.44ecf14d.js:7
ne @ chunk-vendors~d2305125.44ecf14d.js:7
En.t.$emit @ chunk-vendors~d2305125.44ecf14d.js:7
handleMenuItemClick @ chunk-vendors~2a42e354.64b18aae.js:3
ne @ chunk-vendors~d2305125.44ecf14d.js:7
En.t.$emit @ chunk-vendors~d2305125.44ecf14d.js:7
dispatch @ chunk-vendors~2a42e354.64b18aae.js:3
handleClick @ chunk-vendors~2a42e354.64b18aae.js:3
click @ chunk-vendors~2a42e354.64b18aae.js:3
ne @ chunk-vendors~d2305125.44ecf14d.js:7
n @ chunk-vendors~d2305125.44ecf14d.js:7
Qi.o._wrapper @ chunk-vendors~d2305125.44ecf14d.js:7
chunk-vendors~d2305125.44ecf14d.js:7 Error: Request failed with status code 404
    at t.exports (chunk-vendors~2a42e354.64b18aae.js:1)
    at t.exports (chunk-vendors~2a42e354.64b18aae.js:3)
    at XMLHttpRequest.h.onreadystatechange (chunk-vendors~2a42e354.64b18aae.js:3)
```

**Comment by vermorag** [comment]
It looks like you do not have all modules initialized at your backend application. We started the separation of the code parts in 5-7 releases, but this functionality not fully ready for production, so it can cause such errors. The production-ready actual release is 4.4.2.

Please try to run `git submodule update --init`.

**Comment by DigitalLeaves** [comment]
How should I proceed if I downloaded the ZIPs instead of using git clone?

**Comment by vermorag** [comment]
You can download modules as ZIP from our git and unpack them to their folders.
Reports module(Required for reports to work): https://git.amazingcat.net/AmazingTime/core/modules/reports
Gitlab module(Optional): https://git.amazingcat.net/AmazingTime/core/modules/gitlab/backend
Redmine module(Optional): https://git.amazingcat.net/AmazingTime/core/modules/redmine/backend

You should extract them to the `<folder of your backend application>/Modules`. You'll find corresponding folders there.

**Comment by DigitalLeaves** [comment]
Thank you very much @aVyarvelsky. Indeed modules were not installed and now I installed reports, however, I can't make it work.

What I did:
1. I downloaded the ZIP from the URL you told me.
2. I decompressed it in the Modules folder
3. I entered the Reports folder and run 
`composer install`

Still the same error. I might be missing something super obvious, sorry.
The log from Apache2:

`94.26.21.60 - - [25/May/2020:08:07:12 +0000] "POST /api/v1/reports/projects HTTP/1.1" 404 977 "https://myhost.com/report/projects?type=day&start=2020-05-24&end=2020-05-24" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36"`

These are the contents of my Modules folder in the backend.

```
drwxr-xr-x  2 www-data www-data 4096 May 25 06:34 EmailReports
drwxr-xr-x  7 www-data www-data 4096 May 25 06:42 EventListener
drwxr-xr-x 14 www-data www-data 4096 May 25 06:38 GitlabIntegration
drwxr-xr-x  2 www-data www-data 4096 May 25 06:39 Invoices
drwxr-xr-x  2 www-data www-data 4096 May 24 07:32 JiraIntegration
drwxr-xr-x 14 www-data www-data 4096 May 25 06:37 RedmineIntegration
drwxr-xr-x 12 www-data www-data 4096 May 25 06:43 Reports
drwxr-xr-x  2 www-data www-data 4096 May 24 07:32 TrelloIntegration
```

And these the contents of the Reports folder:

```
-rw-r--r--  1 www-data www-data   514 May 24 16:19 composer.json
-rw-r--r--  1 www-data www-data 47219 May 25 06:43 composer.lock
drwxr-xr-x  2 www-data www-data  4096 May 24 16:19 Config
drwxr-xr-x  2 www-data www-data  4096 May 24 16:19 Console
drwxr-xr-x  2 www-data www-data  4096 May 24 16:19 Entities
drwxr-xr-x  3 www-data www-data  4096 May 24 16:19 Exports
drwxr-xr-x  5 www-data www-data  4096 May 24 16:19 Http
-rw-r--r--  1 www-data www-data   253 May 24 16:19 module.json
drwxr-xr-x  2 www-data www-data  4096 May 24 16:19 Providers
drwxr-xr-x  3 www-data www-data  4096 May 24 16:19 Resources
drwxr-xr-x  2 www-data www-data  4096 May 24 16:19 Routes
drwxr-xr-x 17 www-data www-data  4096 May 25 06:43 vendor
```

Can you please help us? 🙏🏻 Thanks again.

**Comment by vermorag** [comment]
Can you please show contents of `modules_statuses.json`? 
Also, you may try run `php artisan config:clear`.

**Comment by DigitalLeaves** [comment]
Yes, there was no modules_statuses.json (just modules_statuses.example.json and modules_statuses.staging.json). I copied modules_statuses.example.json into modules_statuses.json and set them like this:

```
{
  "EmailReports": false,
  "EventListener": true,
  "GitlabIntegration": false,
  "Invoices": false,
  "JiraIntegration": false,
  "RedmineIntegration": true,
  "Reports": true,
  "TrelloIntegration": false
}
```
After that I tried 
`php artisan config:clear`
But still getting error. Now it seems to be a 500 error.

`XX.XX.XX.XX - - [25/May/2020:11:47:29 +0000] "POST /api/v1/projects/list HTTP/1.1" 200 686 "https://mywebhost.com/dashboard/team?type=week&start=2020-05-25&end=2020-05-31" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36"`

`XX.XX.XX.XX - - [25/May/2020:11:47:31 +0000] "POST /api/v1/reports/dashboard HTTP/1.1" 500 438 "https://mywebhost.com/dashboard/team?type=week&start=2020-05-25&end=2020-05-31" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36"`

**Comment by vermorag** [comment]
We moving on. It's progress definitely. =)

Open the console in the browser and go to the "Network" section. 
Try to download the report again and then return back to the Developer tools.
Select request to the `https://myhost.com/api/v1/reports/dashboard` and look at its contents.
Send it to me please.

**Comment by DigitalLeaves** [comment]
Thank you! So this is the console before clicking on generate report:
![Screenshot 2020-05-25 at 18 51 56](https://user-images.githubusercontent.com/5340535/82828372-9427dc00-9eb9-11ea-8c7b-9ce807e8fe6f.jpg)

Then I click on generate the report and get this:
![Screenshot 2020-05-25 at 18 52 11](https://user-images.githubusercontent.com/5340535/82828396-9d18ad80-9eb9-11ea-8209-21cca69c06a4.jpg)

The error in the network request is:
```
{"success":false,"message":"mkdir(): No such file or directory","error_type":"error"}
```
![Screenshot 2020-05-25 at 18 53 25](https://user-images.githubusercontent.com/5340535/82828441-bd486c80-9eb9-11ea-90e1-7a83d377ce49.jpg)

**Comment by vermorag** [comment]
What version of Cattr do you use?

You can check it at the `about` page, or in the composer.json of the backend.

**Comment by DigitalLeaves** [comment]
```
"name": "cattr/backend-application",
"description": "Backend Application for Cattr",
"version": "2.6.0"
```

**Comment by vermorag** [comment]
Hmm... Try to create a folder `<path to backend part>storage/framework/tmp/exporter/` and give write right to it.

It helped for me.

**Comment by DigitalLeaves** [comment]
Cool, now I can download the CSV and XLSX perfectly. The PDF sometimes downloads ok and sometimes downloads at 0 bytes. However, for me at least, I can do with CSV reports. 

If you need information or I can help you debug it, I will be glad to, though. It's the least I can do.

I also sent your team, an email through your website for possible work collaboration, let me know if you didn't receive it. Thanks again!

**Comment by vermorag** [comment]
I've created a ticket for our developers.
You can find it [in our Gitlab](https://git.amazingcat.net/AmazingTime/core/modules/reports/-/issues/1).
I'll keep you updated about it here.

We have received your message and now working on an estimate.

**Comment by vermorag** [comment]
Finally, we released a stable Cattr version 3.0.x where we replaced using of that temporal directory with using of the common Laravel tmp. So it's not necessary to create the dir on your own now.

All should work for now.

**Comment by NasirK84** [comment]
I installed cattr with docker. Reports are not working, I add reports files to Reports folder in Modules and run command: sudo docker exec -it cattr sh -c "/app/backend/Modules/Reports composer install"

But it shows this error: sh: /app/backend/Modules/Reports: Permission denied

2nd after copied Reports files cattr application shows error: Error

Class 'Modules\Reports\Providers\ReportsServiceProvider' not found

Any solution for this.
